# [\#13148 Issue](https://github.com/ethereum/solidity/issues/13148) `closed`: Permit disallowing of fallback function call during `address.call`

#### <img src="https://avatars.githubusercontent.com/u/811305?u=b2a7f87d77c4fd8a388b45691c92912df24ea6e1&v=4" width="50">[lukehutch](https://github.com/lukehutch) opened issue at [2022-06-15 09:47](https://github.com/ethereum/solidity/issues/13148):

## Description

Solidity's fallback function can give rise to [serious security vulnerabilities](https://media.dedaub.com/phantom-functions-and-the-billion-dollar-no-op-c56f062ae49f), potentially resulting in the loss of enormous amounts of money.

Furthermore, the rollover to the fallback function may also give rise to unexpected behavior, e.g. if an ERC777 token recipient registers its `ERC777TokensRecipient` interface properly via ERC1820, but the programmer misspells the notification function `tokensReceived` as `toknsReceived`, and the token recipient contract implements a non-reverting fallback function, the ERC777 token contract will think it has successfully sent to the token recipient contract, even though the `tokensReceived` function is not defined in the token recipient contract. This breaks the ERC777 specification.

There needs to be a way to specify that `.call` should not roll over to the fallback function, e.g.:

`addr.call{value: X, enableFallback: false}( ... );`

Arguably `enableFallback` should even be `false` by default, given the magnitude of the possible vulnerabilities exposed by this.

## Environment

- Compiler version: 0.8.14


#### <img src="https://avatars.githubusercontent.com/u/1347491?v=4" width="50">[ekpyron](https://github.com/ekpyron) commented at [2022-06-15 09:54](https://github.com/ethereum/solidity/issues/13148#issuecomment-1156257021):

And how should this flag work in practice?
In general, you don't know if ``addr`` has a fallback function, resp. with which calldata it will fall back to a fallback function, so I don't see it being possible to do much about this on the call site.

#### <img src="https://avatars.githubusercontent.com/u/1347491?v=4" width="50">[ekpyron](https://github.com/ekpyron) commented at [2022-06-15 09:56](https://github.com/ethereum/solidity/issues/13148#issuecomment-1156259435):

On the callee side, the fallback function is already discouraged in favour of a ``receive`` function (which will revert on non-empty calldata) since Solidity 0.6.0, unless you actually *want* to handle non-empty calldata.

(i.e. if https://library.dedaub.com/contracts/Ethereum/C02AAA39B223FE8D0A0E5C4F27EAD9083C756CC2/source was written against solidity 0.6.0+, it would not have a ``fallback`` function, but a ``receive`` function, which would have prevented the issue. Back in <0.6, it *should* have asserted that calldata is empty, but since people did not reliably do that, we introduced the receive function back then)

#### <img src="https://avatars.githubusercontent.com/u/811305?u=b2a7f87d77c4fd8a388b45691c92912df24ea6e1&v=4" width="50">[lukehutch](https://github.com/lukehutch) commented at [2022-06-15 09:58](https://github.com/ethereum/solidity/issues/13148#issuecomment-1156261197):

Can't the caller verify whether the callee actually implements the required function? I know this can't be done from Solidity per se, but can't it be done somehow in Yul or by the EVM?

#### <img src="https://avatars.githubusercontent.com/u/1347491?v=4" width="50">[ekpyron](https://github.com/ekpyron) commented at [2022-06-15 09:58](https://github.com/ethereum/solidity/issues/13148#issuecomment-1156261715):

No, it can't :-).

#### <img src="https://avatars.githubusercontent.com/u/811305?u=b2a7f87d77c4fd8a388b45691c92912df24ea6e1&v=4" width="50">[lukehutch](https://github.com/lukehutch) commented at [2022-06-15 09:59](https://github.com/ethereum/solidity/issues/13148#issuecomment-1156262307):

This seems like a gaping hole in EVM capability... is there interest in extending the EVM to add ABI introspection?

#### <img src="https://avatars.githubusercontent.com/u/1347491?v=4" width="50">[ekpyron](https://github.com/ekpyron) commented at [2022-06-15 10:01](https://github.com/ethereum/solidity/issues/13148#issuecomment-1156264721):

This would require severe changes to the EVM, e.g. actually implementing a function entry point table instead of being entirely agnostic to interfaces and solely relying on signatures stored in calldata. There are EIPs along those lines IIRC.

#### <img src="https://avatars.githubusercontent.com/u/811305?u=b2a7f87d77c4fd8a388b45691c92912df24ea6e1&v=4" width="50">[lukehutch](https://github.com/lukehutch) commented at [2022-06-15 10:02](https://github.com/ethereum/solidity/issues/13148#issuecomment-1156266094):

Or at least, this could be implemented by the callee, if it is signaled by the caller that the caller does not want the fallback function to ever be called. Then as long as callees are built with a supported version of Solidity, the callee could enforce the caller's fallback preferences. Obviously this still has room for vulnerabilities (anybody could use an older compiler to ignore the caller's wishes), but it would close many loopholes in the majority of cases.

#### <img src="https://avatars.githubusercontent.com/u/1347491?v=4" width="50">[ekpyron](https://github.com/ekpyron) commented at [2022-06-15 10:03](https://github.com/ethereum/solidity/issues/13148#issuecomment-1156267445):

But yeah, from the Solidity side, currently, this can only really be addressed at the callee-side, which was done in 0.6.0 with introducing the ``receive`` function as alternative to a full blown ``fallback``.

The way the caller signals the callee that it wants the ``receive`` function and not the ``fallback`` function, is to use empty calldata.

#### <img src="https://avatars.githubusercontent.com/u/811305?u=b2a7f87d77c4fd8a388b45691c92912df24ea6e1&v=4" width="50">[lukehutch](https://github.com/lukehutch) commented at [2022-06-15 10:05](https://github.com/ethereum/solidity/issues/13148#issuecomment-1156268762):

Wouldn't a zero-parameter payable function rollover to `receive` then, because the calldata would be empty?  (I have one of these functions in my code.) That's the same vulnerability, because the wrong function may get called if the function name has a typo or something.

#### <img src="https://avatars.githubusercontent.com/u/1347491?v=4" width="50">[ekpyron](https://github.com/ekpyron) commented at [2022-06-15 10:06](https://github.com/ethereum/solidity/issues/13148#issuecomment-1156270513):

I.e. these days what *should* happen is:

- The callee should usually just implement a ``receive`` function, which will only be called with empty calldata. Even if it additionally implements a ``fallback`` function, empty calldata will indicate that the ``receive`` function is desired.
- The caller just uses empty calldata to indicate that it does not want to call a function.

So in this sense this is solved by Solidity 0.6.0.



> Wouldn't a zero-parameter payable function rollover to `receive` then, because the calldata would be empty? (I have one of these functions in my code.) That's the same vulnerability, because the wrong function may get called if the function name has a typo or something.

The function signature makes up the first four bytes of calldata, so if you attempt to call an actual function, you never have empty calldata.

#### <img src="https://avatars.githubusercontent.com/u/811305?u=b2a7f87d77c4fd8a388b45691c92912df24ea6e1&v=4" width="50">[lukehutch](https://github.com/lukehutch) commented at [2022-06-15 10:11](https://github.com/ethereum/solidity/issues/13148#issuecomment-1156274998):

Ah, got it. Thanks for taking the time to explain.

Well I can already fix this on the callee side, to protect the callee contract from causing problems for other contracts at least, by adding a definition for `fallback` that simply reverts. But it's the caller side I'm worried about. In the URL I provided, the caller has no way to know that their call failed, which creates the security vulnerability. In the ERC777 recipient notification example I gave, Solidity's automatic fallback call actually breaks the ERC777 spec.

#### <img src="https://avatars.githubusercontent.com/u/1347491?v=4" width="50">[ekpyron](https://github.com/ekpyron) commented at [2022-06-15 10:23](https://github.com/ethereum/solidity/issues/13148#issuecomment-1156287417):

Yes, prior to 0.6.0, the callee's ``fallback`` function was supposed to consider whether it gets calldata or not on its own... so https://library.dedaub.com/contracts/Ethereum/C02AAA39B223FE8D0A0E5C4F27EAD9083C756CC2/source?line=35 should, instead of
```
    function() public payable {
        deposit();
    }
```
have had
```
    function() public payable {
        require(msg.data.length == 0, "I only want to receive ether, but I got non-empty calldata!");
        deposit();
    }
```

But since this is missing in cases like this, in 0.6.0 we split this on the language level into ``receive()`` and ``fallback()``, which makes the distinction clearer (if you don't define a ``fallback()``, but only a ``receive()``, any calls to unknown functions will automatically revert).

But there's nothing that can be done by Solidity about existing, deployed contracts that are, in this sense, buggy :-).

#### <img src="https://avatars.githubusercontent.com/u/811305?u=b2a7f87d77c4fd8a388b45691c92912df24ea6e1&v=4" width="50">[lukehutch](https://github.com/lukehutch) commented at [2022-06-15 10:39](https://github.com/ethereum/solidity/issues/13148#issuecomment-1156307564):

I read the documentation I could find on `fallback()` and `receive()`, including the 0.6.0 post about this. But I'm still hazy on what happens if these are not defined. So if I don't define `fallback()` or `receive()`, but (for argument's sake) I have a `payable` function of a different name in my contract, then I can only call the contract with functions that are actually defined (because there is no `fallback()`, I can't send ETH directly to the contract address using `.send()` or `.transfer()`, but I can send ETH to the `payable` function that I defined that is not `receive`?

I'm trying to determine if there is any reason to add this code (or is this exactly what happens if these functions are not defined?).

```
    fallback() external {
        revert();
    }

    receive() external payable {
        revert();
    }
```

#### <img src="https://avatars.githubusercontent.com/u/1347491?v=4" width="50">[ekpyron](https://github.com/ekpyron) commented at [2022-06-15 10:54](https://github.com/ethereum/solidity/issues/13148#issuecomment-1156320633):

There's no reason to add this code. If you neither define a ``fallback``, nor a ``receive`` function, then calls to non-existing functions will automatically revert.
Even if you just don't define a ``fallback`` function and only a ``receive`` function, *still* all calls to non-existing will automatically revert (only calls with no calldata, including no function signature, will pass to the ``receive`` function).

#### <img src="https://avatars.githubusercontent.com/u/811305?u=b2a7f87d77c4fd8a388b45691c92912df24ea6e1&v=4" width="50">[lukehutch](https://github.com/lukehutch) commented at [2022-06-15 11:17](https://github.com/ethereum/solidity/issues/13148#issuecomment-1156343146):

OK, thanks.

If there really is no way for the caller to prevent calls with non-empty calldata triggering the fallback function, then you can go ahead and close this issue (although it's a real shame that this is not possible...).


-------------------------------------------------------------------------------



[Export of Github issue for [ethereum/solidity](https://github.com/ethereum/solidity). Generated on 2024.12.15 at 06:45:24.]
